architecture/ambient/lifecycle.md

```diff
@@ -0,0 +1,136 @@
+The CNI plugin synchronously runs before application containers start, ensure we can setup the network configuration and start Ztunnel.
+Only once those two conditions are met do we mark the CNI plugin as successfully complete.
+
+From the above we are ensured that when the application starts and opens a connection, Ztunnel will be ready to `accept()` it.
+However, this is not enough to fully serve the request - Ztunnel will need the certificate (from the CA) and full workload information (from the XDS server).
+This is done through a waiting mechanism: Ztunnel will accept all connections, but it will not start forwarding traffic until all the above is fetched.
+This "lazy loading" optimizes pod startup time, for the common case where an application is not sending outbound traffic instantly.
+For applications that do, they will see a slight latency increase during startup; most applications can handle this far better than failed connections.
+
+## Pod Shutdown
+
+### Pod Shutdown Requirements
+
+During Pod shutdown, we MUST meet the constraint that the Pod shutting down can send traffic up until its full stopped.
+Pod shutdown can be a long process; while `terminationGracePeriodSeconds` is typically ~30s, it can be extremely high, leaving terminating pods running for extended periods of time with an expectation of networking.
+
+Additionally, we MUST accept incoming traffic during this time, both new and existing.
+Typically, it is rare to get new connections after shutdown has started - once shutdown is triggered the endpoint is marked `NotReady` and removed from load balancing consideration.
+However, some applications do not utilize Service readiness (connecting directly to the pod, etc).
+Additionally, eventually consistency means we will often get some final requests trickle in before all clients are aware the Pod is shutting down.
+
+Finally, we SHOULD tell our (HBONE) peers we are shutting down.
+This can help ensure that clients do not hold onto pooled connections for too long.
+This is not strictly required; Ztunnel clients have both keepalives (which will start to fail once the other end is shut down) and pool idle timeouts (which will close the connection when it has had no activity for some period; because the backend is gone, there will be no activity).
+However, it would be ideal to give immediate feedback rather than relying on these mechanisms.
+
+### Pod Shutdown Implementation
+
+[ZDS](../../pkg/zdsapi/zds.proto) exposes a `DelWorkload` message to signal to Ztunnel that a pod has been terminated.
+This can be triggered either by a Pod being actually deleted, or has transitioned to a [terminal phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/).
+At this point, the pod and all containers have fully shut down, and we can tear everything down.
+
+Because the application(s) are already entirely gone, there is no need to have any "draining" time -- we can immediately terminate.
```

Comment:
Note: there is no need, but at the time of writing we do drain. I willwork on fixing it.

Reply:
xref: https://github.com/istio/ztunnel/issues/899